### PR TITLE
Adding deprecation warning for Flash Attention 1 and user warning against using Triton attention.

### PR DESCRIPTION
**Pull request by ShashankMosaicML**

### CODE REVIEW

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 1]
```diff
@@ -222,6 +223,20 @@ def _validate_config(self) -> None:
                 'attn_impl'] not in ['torch', 'triton']:
             raise NotImplementedError(
                 'prefix_lm only implemented with torch and triton attention.')
+
+        if self.attn_config['attn_impl'] == 'flash' and is_flash_v1_installed():
+            warnings.warn(
+                DeprecationWarning(
+                    'The support for Flash Attention v1 will soon be deprecated. Please upgrade to Flash Attention v2.4.2.'
```
Comment (vchiley) [inline llmfoundry/models/mpt/configuration_mpt.py:230]:
```suggestion
                    'Support for Flash Attention v1 is deprecated. Please upgrade to Flash Attention >= v2.4.2.'
```

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 2]
```diff
@@ -222,6 +223,20 @@ def _validate_config(self) -> None:
...
             raise NotImplementedError(
                 'prefix_lm only implemented with torch and triton attention.')
+
+        if self.attn_config['attn_impl'] == 'flash' and is_flash_v1_installed():
+            warnings.warn(
+                DeprecationWarning(
+                    'Support for Flash Attention v1 is deprecated. Please upgrade to Flash Attention v2.4.2.'
+                ))
+
+        if self.attn_config[
```
Comment (dakinggg) [inline llmfoundry/models/mpt/configuration_mpt.py:233]: does prefix lm require triton?

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 3]
```diff
@@ -222,6 +223,20 @@ def _validate_config(self) -> None:
                 'attn_impl'] not in ['torch', 'triton']:
             raise NotImplementedError(
                 'prefix_lm only implemented with torch and triton attention.')
+
+        if self.attn_config['attn_impl'] == 'flash' and is_flash_v1_installed():
```
Comment (dakinggg) [inline llmfoundry/models/mpt/configuration_mpt.py:227]: add the instructions for how to install and what will change in future version. so, say, you can install with .[gpu-flash2], and after deprecation .[gpu] will be updated to install flash2

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 2 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py:233]: Yes, we need triton for prefix lm because flash attention does not support prefix lm right now.

Comment (dakinggg) [reply llmfoundry/models/mpt/configuration_mpt.py:233]: 😢

Comment (vchiley) [reply llmfoundry/models/mpt/configuration_mpt.py:233]: 😢

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 3 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py:227]: Added the instruction on how to install Flash 2 in the warning, and updated readme to say that after we remove support for FA v1, .[gpu] will be updated to install flash2.

**Review by dakinggg — COMMENTED** [review]
Could you also update the training readme that explains how to use flash attention?

**Comment by ShashankMosaicML** [comment]
> Could you also update the training readme that explains how to use flash attention?

Done. Please take a look.

**Review by vchiley — APPROVED** [review]

**Review by dakinggg — APPROVED** [review]

**Pull request merged**